src/plugins/part.c

```diff
@@ -582,6 +591,136 @@ static BDPartSpec* get_part_spec (PedDevice *dev, PedDisk *disk, PedPartition *p
     return ret;
 }
 
+static BDPartFlag get_flag_from_guid (const gchar *guid) {
+    guint last_flag = 0;
+
+    if (!guid)
+        return BD_PART_FLAG_BASIC_LAST;
+
+    last_flag = log2i (BD_PART_FLAG_BASIC_LAST);
+
+    for (guint i = 1; i < last_flag; i++) {
+      if (!part_flags[i - 1].guid)
+          continue;
+
+      if (g_ascii_strcasecmp (part_flags[i - 1].guid, guid) == 0)
+          return 1 << i;
+    }
+
+    return BD_PART_FLAG_BASIC_LAST;
+}
+
+static BDPartFlag get_flag_from_id (guint id) {
+    gchar *id_str = NULL;
+    guint last_flag = 0;
+
+    id_str = g_strdup_printf ("0x%.2x", id);
+    last_flag = log2i (BD_PART_FLAG_BASIC_LAST);
+
+    for (guint i = 1; i < last_flag; i++) {
+        if (!part_flags[i - 1].id)
+            continue;
+
+        if (g_strcmp0 (part_flags[i - 1].id, id_str) == 0) {
+            g_free (id_str);
+            return 1 << i;
+        }
+    }
+
+    g_free (id_str);
+    return BD_PART_FLAG_BASIC_LAST;
+}
+
+static BDPartSpec* get_part_spec_fdisk (struct fdisk_context *cxt, struct fdisk_partition *pa, GError **error) {
+    struct fdisk_label *lb = NULL;
+    struct fdisk_parttype *ptype = NULL;
+    guint part_id = 0;
+    BDPartSpec *ret = NULL;
+    const gchar *devname = NULL;
+    const gchar *partname = NULL;
+    BDPartFlag bd_flag;
+
+    ret = g_new0 (BDPartSpec, 1);
+
+    devname = fdisk_get_devname (cxt);
+
+    if (fdisk_partition_has_partno (pa)) {
+        if (isdigit (devname[strlen(devname) - 1]))
+            ret->path = g_strdup_printf ("%sp%zu", devname, fdisk_partition_get_partno (pa) + 1);
+        else
+            ret->path = g_strdup_printf ("%s%zu", devname, fdisk_partition_get_partno (pa) + 1);
+    }
+
+    partname = fdisk_partition_get_name (pa);
+    if (partname)
+        ret->name = g_strdup (partname);
+
+    if (fdisk_partition_is_container (pa))
+        ret->type = BD_PART_TYPE_EXTENDED;
+    else if (fdisk_partition_is_nested (pa))
+        ret->type = BD_PART_TYPE_LOGICAL;
+    else
+        ret->type = BD_PART_TYPE_NORMAL;
+
+    if (fdisk_partition_is_freespace (pa))
+        ret->type |= BD_PART_TYPE_FREESPACE;
+
+    if (fdisk_partition_has_start (pa))
+        ret->start = (guint64) fdisk_partition_get_start (pa) * fdisk_get_sector_size (cxt);
+
+    if (fdisk_partition_has_size (pa))
+        ret->size = (guint64) fdisk_partition_get_size (pa) * fdisk_get_sector_size (cxt);
+
+    lb = fdisk_get_label (cxt, NULL);
+    if (!lb) {
+        g_set_error (error, BD_PART_ERROR, BD_PART_ERROR_FAIL,
+                     "Failed to read partition table.");
+        bd_part_spec_free (ret);
+        return NULL;
+    }
+
+    if (g_strcmp0 (fdisk_label_get_name (lb), "gpt") == 0 && fdisk_partition_has_partno (pa)) {
```

Comment:
What's the reason for the `has_partno()` check?

Reply:
I'm using `get_partno` below. I could probably just use `ret->type == BD_PART_TYPE_NORMAL` because only "normal" partitions have GUIDs.

Reply:
Both are equally cryptic without a comment. Maybe split this into two `if`s because otherwise this *either DOS or GPT-with-part-number* looks quite confusing, making me think something should be done for GPT partitions with no number (which I cannot really imagine).

Reply:
Ok, makes sense. I've added a separate `if` and a comment.